proposals/0000-mutable-fields.rst

```diff
@@ -0,0 +1,473 @@
+
+To reduce the overhead of the existing ``IORef``, a common technique
+is to use ``{-# UNPACK #-}`` annotations on ``IORef`` fields in
+constructors. We also rely on the strictness analyser to eliminate the
+``IORef`` box whenever possible.  Neither of these techniques would
+work with the new ``IORef``, which cannot be unpacked. However the new
+``IORef`` has the equivalent representation to an unboxed old
+``IORef``, so performance will usually be the same.
+
+There's one respect in which performance of the new ``IORef`` type
+might not be identical to that of an unpacked old ``IORef``
+(i.e. ``MutVar#``): ``IORef`` is lifted, whereas ``MutVar#`` is not;
+so a strict ``IORef`` may still need to be evaluated, depending on
+whether we can tell from the context that the ``IORef`` is already
+evaluated.  There's a way to mitigate this effect: we can provide::
+
+  ioRefMutable :: IORef a -> Mutable a
+
+the ``Mutable a`` doesn't need to be evaluated.
+
+
+Mutable unboxed fields
+~~~~~~~~~~~~~~~~~~~~~~
+
+We can extend this to handle mutable unboxed fields too::
+
+  Ref# :: forall r. * -> TYPE 'r -> Type Ref
+
+It's perfectly fine for ``Ref#`` to be parameterised by a
+representation-polymorphic type because the representation of ``Ref#``
+itself does not depend on this type argument.
+
+However, we now need a family of primitives to work with these::
+
+  readRefInt#  :: Ref# s Int# -> State# s -> (# State# s, Int# #)
+  writeRefInt# :: Ref# s Int# -> Int# -> State# s -> (# State# s, () #)
+
+  readRefDouble#  :: Ref# s Double# -> State# s -> (# State# s, Double# #)
+  writeRefDouble# :: Ref# s Double# -> Double# -> State# s -> (# State# s, () #)
+
+and so on.
+
+Deriving
+~~~~~~~~
+
+A type with one or more mutable constructors can derive only ``Eq`` and
+`Typeable`.
+
+`Eq` is supported by using ``reallyUnsafePtrEquality#`` to compare
+mutable constructors, but we must ensure that the constructors are
+evaluated strictly in the same way as we do for ``dataToTag#``.
+
+
+Mutable constructors with no mutable fields
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+We also propose to make it possible to declare a mutable constructor
+without any mutable fields, for example::
+
+ data IdPair :: * -> * -> * where
+     IdPair :: a -> b -> IO (IdPair a b)
+
+The monadic return type indicates that ``IdPair`` behaves as a mutable
+constructor, in that:
+
+- Its constructor has the declared type
+- It has identity, and equality is implemented using pointer equality
+  (see "Deriving" above).
+
+The degenerate example is a nullary mutable constructor::
+
+ data Token where
+   Token :: IO Token
+   deriving Eq
+
+Each instance of ``Token`` must be unique, because the ``IO``
+operation creates a new one that should compare equal only to itself.
+Therefore the usual optimisation for nullary constructors whereby we
+share a single global copy of the constructor does *not* apply to
+mutable constructors, and we must always allocate them in the heap.
```

Comment:
We have `Unique` for that use case. It performs better and is less quirky wrt. evaluatedness.

If you want to somehow optimise `Eq` instances for pointer equality, I'd much rather support a design like https://arxiv.org/pdf/2003.01685.pdf. E.g., export a primitive

```hs
-- | Semantically equivalent to `\a b k -> k`, but GHC is free to return `True`
-- if `a` and `b` are referentially equal and omit evaluating `k`.
GHC.Exts.withPtrEq :: a -> a -> Bool -> Bool

instance Eq (MutPair a b) where
  MutPair a1 b1 == MutPair a2 b2 = withPtrEq a1 a2 (a1 == a2) && withPtrEq b1 b2 (b1 == b2)
```